runtime/browser/devtools/xwalk_devtools_frontend.cc

```diff
@@ -0,0 +1,345 @@
+                 shell_devtools_, "DevToolsAPI.streamWrite",
+                 base::Owned(id), base::Owned(chunk), nullptr));
+  return num_bytes;
+}
+
+int ResponseWriter::Finish(const net::CompletionCallback& callback) {
+  return net::OK;
+}
+
+}  // namespace
+
+// This constant should be in sync with
+// the constant at devtools_ui_bindings.cc.
+const size_t kMaxMessageChunkSize = IPC::Channel::kMaximumMessageSize / 4;
+
+// static
+XWalkDevToolsFrontend* XWalkDevToolsFrontend::Show(
+    content::WebContents* inspected_contents) {
+  Runtime* runtime = Runtime::Create(static_cast<XWalkBrowserContext*>(
+      inspected_contents->GetBrowserContext()));
+  runtime->set_ui_delegate(DefaultRuntimeUIDelegate::Create(runtime));
+  runtime->Show();
+  XWalkDevToolsFrontend* devtools_frontend = new XWalkDevToolsFrontend(
+      static_cast<NativeAppWindowDesktop*>(runtime->window()),
+      inspected_contents);
+  runtime->set_observer(devtools_frontend);
+  devtools_http_handler::DevToolsHttpHandler* http_handler =
+      XWalkContentBrowserClient::Get()->main_parts()
+          ->devtools_http_handler();
+  runtime->LoadURL(http_handler->GetFrontendURL("/devtools/inspector.html"));
+
+  return devtools_frontend;
+}
+
+void XWalkDevToolsFrontend::Activate() {
+  runtime_window_->ActivateContents(web_contents());
+}
+
+void XWalkDevToolsFrontend::Focus() {
+  web_contents()->Focus();
+}
+
+void XWalkDevToolsFrontend::InspectElementAt(int x, int y) {
+  if (agent_host_)
+    agent_host_->InspectElement(x, y);
+}
+
+void XWalkDevToolsFrontend::Close() {
+  runtime_window_->Close();
+}
+
+void XWalkDevToolsFrontend::DisconnectFromTarget() {
+  if (!agent_host_)
+    return;
+  agent_host_->DetachClient();
+  agent_host_ = NULL;
+}
+
+void XWalkDevToolsFrontend::OnRuntimeClosed(Runtime* runtime) {
+  delete runtime;
+}
+
+XWalkDevToolsFrontend::XWalkDevToolsFrontend(
+    NativeAppWindowDesktop* runtime_window,
+    content::WebContents* inspected_contents)
+    : content::WebContentsObserver(runtime_window->web_contents()),
+      runtime_window_(runtime_window),
+      inspected_contents_(inspected_contents),
+      weak_factory_(this) {
+}
+
+XWalkDevToolsFrontend::~XWalkDevToolsFrontend() {
+  for (const auto& pair : pending_requests_)
+    delete pair.first;
+}
+
+void XWalkDevToolsFrontend::RenderViewCreated(
+    content::RenderViewHost* render_view_host) {
+  if (!frontend_host_) {
+    frontend_host_.reset(content::DevToolsFrontendHost::Create(
+        web_contents()->GetMainFrame(), this));
+  }
+}
+
+void XWalkDevToolsFrontend::DocumentAvailableInMainFrame() {
+  agent_host_ = content::DevToolsAgentHost::GetOrCreateFor(inspected_contents_);
+  agent_host_->AttachClient(this);
+}
+
+void XWalkDevToolsFrontend::WebContentsDestroyed() {
+  if (agent_host_)
+    agent_host_->DetachClient();
+  delete this;
+}
+
+void XWalkDevToolsFrontend::HandleMessageFromDevToolsFrontend(
+    const std::string& message) {
+  if (!agent_host_)
```

Comment:
override of content::DevToolsFrontendHost::Delegate implementation if it's what you're asking. Pure virtual method, needs to be reimplemented.